projects/pgai/pgai/vectorizer/embedders/openai.py

```diff
@@ -150,6 +150,18 @@ async def call_embed_api(self, documents: list[str]) -> EmbeddingResponse:
             embeddings=embeddings, usage=Usage(prompt_tokens, total_tokens)
         )
 
+    def _estimate_token_length(self, document: str) -> float:
+        """
+        Estimates token count based on UTF-8 byte length.
+        """
+
+        total_estimated_tokens = 0
+        for char in document:
+            byte_length = len(char.encode("utf-8"))
+            total_estimated_tokens += byte_length * 0.25  # 0.25 tokens per byte
+
+        return total_estimated_tokens
+
```

Comment:
I would suggest that we do this, and then remove the type changes that this propagates:

```suggestion
    def _estimate_token_length(self, document: str) -> int:
        """
        Estimates token count based on UTF-8 byte length.
        """

        total_estimated_tokens = 0
        for char in document:
            byte_length = len(char.encode("utf-8"))
            total_estimated_tokens += byte_length * 0.25  # 0.25 tokens per byte

        return ceil(total_estimated_tokens)
```

Reply:
I don't think that works because that would mean it rounds each individual documents tokens not the whole batch. The api however counts the bytes of everything first and then rounds.

E.g. if you send 100 batches of just "a" it would be estimated with 100 tokens this way but the api only assigns 25 (which makes no sense but well that's how it works)

Reply:
Hmmm... interesting. I didn't realise that on first inspection.